File: solutions/typescript/problems/[0023_Hard] Merge k Sorted Lists.ts
// Definition for singly-linked list.
// @ts-ignore
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    if (lists.length === 0) return null;
    return lists.reduce((l1, l2) => mergeTwoLists(l1, l2));
}

function mergeTwoLists(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const pointerToHead = new ListNode();
    let currentNode = pointerToHead;
    while (l1 || l2) {
        if (l1 && l2) {
            if (l1.val <= l2.val) {
                currentNode.next = l1;
                currentNode = currentNode.next;
                l1 = l1.next;
            } else {
                currentNode.next = l2;
                currentNode = currentNode.next;
                l2 = l2.next;
            }
        } else if (l1) {
            currentNode.next = l1;
            currentNode = currentNode.next;
            l1 = l1.next;
        } else if (l2) {
            currentNode.next = l2;
            currentNode = currentNode.next;
            l2 = l2.next;
        }
    }
    return pointerToHead.next;
}

File: solutions/c++/problems/[0023_Hard] Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct compare {
    bool operator()(ListNode* &node1, ListNode* &node2) {
        return node1->val > node2->val;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* &node1, ListNode* &node2) {return node1->val < node2->val;};
        priority_queue<ListNode*, vector<ListNode*>, compare> p;
        for(auto list: lists) {
            if(list) p.push(list);
        }
        ListNode* toHead = new ListNode();
        auto curr = toHead;
        while(!p.empty()) {
            curr->next = p.top();
            p.pop();
            if (curr->next->next) p.push(curr->next->next);
            curr = curr->next;
        }
        return toHead->next;
    }
};